### PR TITLE
New unresolver implementation

### PR DESCRIPTION
**Pull request by stsewd**
- Move unresolve_domain to unresolver
- Use the new implementation from https://github.com/readthedocs/readthedocs.org/pull/9425  for the unresolver
- Remove settings override from the unresolver, we don't need it.

This doesn't use this new implementation to handle the URLs from proxito nor has support for a custom urlconf, those things will be implemented in another PR.

This is on top of https://github.com/readthedocs/readthedocs.org/pull/9462

----
:books: Documentation preview :books:: https://docs--9500.org.readthedocs.build/en/9500/

----
:books: Documentation preview :books:: https://dev--9500.org.readthedocs.build/en/9500/

### CODE REVIEW

File: readthedocs/core/unresolver.py [hunk 1]
```diff
@@ -1,99 +1,279 @@
...
-            version_slug=version_slug,
-            filename=kwargs.get('filename', ''),
-        )
+        By default any path will match.
+        """
+        file = self._normalize_filename(path)
+        version = canonical_project.versions.filter(
+            slug=canonical_project.default_version
+        ).first()
+        if version:
```
Comment (ericholscher) [inline readthedocs/core/unresolver.py:171]: In what case will this be `False`? I think when they have a default_version that doesn't exist?

**Review by ericholscher — COMMENTED** [review]
This logic makes a good bit of sense. I want to better understand the whole picture of this work. Is the next step to use the unresolver in proxito to standardize how we're parsing this in each place in the code?

File: readthedocs/core/unresolver.py [hunk 2]
```diff
@@ -1,99 +1,279 @@
...
-        parsed = urlparse(path)
-        path = parsed.path
-        project_slug = getattr(request, 'host_project_slug', None)
+        Try to match a subproject.
 
-        if not project_slug:
+        If the subproject exists, we try to resolve the rest of the path
+        with the subproject as the canonical project.
+        """
+        match = self.subproject_pattern.match(path)
```
Comment (ericholscher) [inline readthedocs/core/unresolver.py:142]:
```suggestion
        #  TODO: Allow the canonical_project to override this url somehow..
        # if canonical_project.subproject_pattern:
        #     match = canonical_project.subproject_pattern.match(path)
        match = self.subproject_pattern.match(path)
```

Is this the longer term goal for customization?

Comment (stsewd) [reply readthedocs/core/unresolver.py:142]: yeah, basically on each call to match, will be some extra logic to use the custom pattern from the project instead of the default one.

File: readthedocs/core/unresolver.py (hunk 1 above)
Comment (stsewd) [reply readthedocs/core/unresolver.py:171]: yeah, this is when a use doesn't have a default version or the default version is invalid

File: readthedocs/core/unresolver.py [hunk 3]
```diff
@@ -1,99 +1,279 @@
...
-    'Unresolved', 'project, language_slug, version_slug, filename, fragment')
+
+@dataclass(slots=True)
+class UnresolvedURL:
+
+    """Dataclass with the parts of an unresolved URL."""
+
+    # This is the project that owns the domain,
+    # this usually is the parent project of a translation or subproject.
+    canonical_project: Project
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:22]: +1 on parent project

**Comment by stsewd** [comment]
> Is the next step to use the unresolver in proxito to standardize how we're parsing this in each place in the code?

Yes, proxito will eventually use this to serve the documentation, but first I'll add the support for custom urlconfs (or a port of that), since we have users using that feature already on .com.

**Review by ericholscher — APPROVED** [review]
I think these changes look good. @humitos not sure if you wanted to review this.

@stsewd This is safe to merge and use before we implement it in proxito right? If so, I'm fine with it going on this deploy.

**Comment by stsewd** [comment]
> @stsewd This is safe to merge and use before we implement it in proxito right? If so, I'm fine with it going on this deploy.

Yeah, it's safe to deploy.

**Comment by ericholscher** [comment]
@stsewd Great, I'd say let's get it merged then so we can do dev on it this week to catch any outstanding issues.

**Comment by humitos** [comment]
@ericholscher 

> I think these changes look good. @humitos not sure if you wanted to review this.

Yeah, I want to review this still. I got pretty overload of things and wasn't able to jump into this yet.

I'll try to do it tomorrow.

File: readthedocs/core/unresolver.py [hunk 4]
```diff
@@ -1,99 +1,284 @@
...
+        If the returned project is `None`, then we weren't able to
+        unresolve the path into a project.
+
+        If the returned version is `None`, then we weren't able to
+        unresolve the path into a valid version of the project.
+
+        :param parent_project: The project that owns the path.
+        :param path: The path to unresolve.
+        :param check_subprojects: If we should check for subprojects,
+         this is used to call this function recursively.
```
Comment (humitos) [inline readthedocs/core/unresolver.py:188]: This could expand about how it's used.

It seems we are calling this recursively _only_ for the subproject case. I guess this is because we want to know if the subproject is multiversion or single version, but it can't have subprojects since we don't support "subprojects on subprojects"

File: readthedocs/core/unresolver.py [hunk 5]
```diff
@@ -1,99 +1,284 @@
...
+
+    # TODO: make this a private method once
+    # proxito uses the unresolve method directly.
+    def unresolve_domain(self, domain):
+        """
+        Unresolve domain by extracting relevant information from it.
+
+        :param str domain: Domain to extract the information from.
+        :returns: A tuple with: the project slug, domain object, and if the domain
+         is from an external version.
```
Comment (humitos) [inline readthedocs/core/unresolver.py:238]: It would be good to make this a dataclass or similar. Returning these structured tuples is always hard to parse mentally and keep that in mind while working with this code and switching between files

File: readthedocs/core/unresolver.py [hunk 6]
```diff
@@ -1,99 +1,284 @@
...
+
+        If the returned version is `None`, then we weren't able to
+        unresolve the path into a valid version of the project.
+
+        :param parent_project: The project that owns the path.
+        :param path: The path to unresolve.
+        :param check_subprojects: If we should check for subprojects,
+         this is used to call this function recursively.
+
+        :returns: A tuple with: project, version, and file name.
```
Comment (humitos) [inline readthedocs/core/unresolver.py:190]: It would be good to make this a dataclass or similar here as well and all these related methods.

File: readthedocs/core/unresolver.py [hunk 7]
```diff
@@ -1,99 +1,284 @@
...
+         is from an external version.
+        """
+        public_domain = self.get_domain_from_host(settings.PUBLIC_DOMAIN)
+        external_domain = self.get_domain_from_host(
+            settings.RTD_EXTERNAL_VERSION_DOMAIN
         )
-        return UnresolvedObject(final_project, lang_slug, version_slug, filename, parsed.fragment)
 
+        subdomain, *root_domain = domain.split(".", maxsplit=1)
+        root_domain = root_domain[0] if root_domain else ""
```
Comment (humitos) [inline readthedocs/core/unresolver.py:246]: I think it's not required use `*root_domain` here. Why are you converting it to a list of one element? Since `maxsplit=1` it will always return a list of 2 elements here.

```
In [10]: subdomain, root_domain = 'docs.readthedocs.io'.split('.', maxsplit=1)

In [11]: subdomain
Out[11]: 'docs'

In [12]: root_domain
Out[12]: 'readthedocs.io'
```

File: readthedocs/core/unresolver.py [hunk 8]
```diff
@@ -1,99 +1,284 @@
...
+            # Serve from the PUBLIC_DOMAIN, ensuring it looks like `foo.PUBLIC_DOMAIN`.
+            if public_domain == root_domain:
+                project_slug = subdomain
+                log.info("Public domain.", domain=domain)
+                return project_slug, None, False
 
-class Unresolver(SettingsOverrideObject):
+            # TODO: This can catch some possibly valid domains (docs.readthedocs.io.com)
+            # for example, but these might be phishing, so let's ignore them for now.
+            log.warning("Weird variation of our domain.", domain=domain)
```
Comment (humitos) [inline readthedocs/core/unresolver.py:257]: I took a look at this at New Relic and it seems we don't have any log for this: https://onenr.io/0oQD4oYZ5Qy 👍🏼

File: readthedocs/core/unresolver.py [hunk 9]
```diff
@@ -1,99 +1,284 @@
...
+        domain_object = (
+            Domain.objects.filter(domain=domain).prefetch_related("project").first()
+        )
+        if domain_object:
+            log.info("Custom domain.", domain=domain)
+            project_slug = domain_object.project.slug
+            return project_slug, domain_object, False
 
-    _default_class = UnresolverBase
+        log.info("Invalid domain.", domain=domain)
```
Comment (humitos) [inline readthedocs/core/unresolver.py:279]: We do have some of these on New Relic: https://onenr.io/0Zw06DrADjv

**Review by humitos — COMMENTED** [review]
This looks good! 💯 

I added some comments about styles and naming. My idea here is to keep consistency through all the codebase and make the names pretty explicit. I think it's important to read the name of the variable and immediately realize what its refers to. As example, `subproject_alias` works better than `project_slug`; since `project_slug` is a lot more generic and could refer to something different than the subproject's alias.

Also, I'd like see those returned tuples converted into dataclasses or similar objects that we can access them via dot notation with clear names.

It's also worth to note that this implementation supports the following cases:

- single version
- multi-language multi-version
- subprojects with single version
- subprojects with multi-language multi-version

However, it does not support "multi-language single-version".

**Comment by humitos** [comment]
> I added some comments about styles and naming. My idea here is to keep consistency through all the codebase and make the names pretty explicit. 

Also, I just noticed that we are calling `root_project` in the design doc, but in the implementation we are just calling it `project`. I think it's important to keep consistency here so we all speak the same language and refer to the same thing with the same name. It seems that `root_project` is more specific and I think it makes sense to use it in the code as well: https://dev.readthedocs.io/en/stable/design/better-doc-urls-handling.html#alternative-implementation

File: readthedocs/core/unresolver.py [hunk 10]
```diff
@@ -1,99 +1,284 @@
...
+        project_slug = match.group("project")
+        file = self._normalize_filename(match.group("file"))
+        project_relationship = (
+            parent_project.subprojects.filter(alias=project_slug)
+            .prefetch_related("child")
+            .first()
         )
+        if project_relationship:
+            return self._unresolve_path(
+                parent_project=project_relationship.child,
```
[the remark this answers is not in the document]
Comment (stsewd) [reply readthedocs/core/unresolver.py:155]: The subproject would be the new parent project, so we can recursively resolve its translations, etc.

File: readthedocs/core/unresolver.py (hunk 7 above)
Comment (stsewd) [reply readthedocs/core/unresolver.py:246]: It returns only one element if the string doesn't have a `.`.

File: readthedocs/core/unresolver.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/core/unresolver.py:190]: I think it is okay to have this as a tuple, they are internal methods, and we already have one dataclass for the final result.

Comment (humitos) [reply readthedocs/core/unresolver.py:190]: Even if they are internal methods, real objects are a lot easier to read and follow while debugging/reading this code.
